### ISSUE TITLE
[GLPI 10.0.5] FC 2.13.2 Problem Attachments

### ISSUE THREAD

**Issue by Gitnor77** [pluginsGLPI/formcreator#3076]
Hi,
I have the same problem #2871  with [GLPI 10.0.5] and [formcreator 2.13.2].

Fix is implemented:
![image](https://user-images.githubusercontent.com/109066311/201018853-b894bbd9-4960-4c09-ac5f-cf401181e2b4.png)

**Comment by Gitnor77** [comment]
Hi,
Any update?
I have a problem with adding an attachment to a created form, the same as previously.
The form itself must be public (accessible to non-registered users). See below PrtScr, please.
![image](https://user-images.githubusercontent.com/109066311/201581041-e8e62de0-6537-4702-a494-1e68fea9e1d9.png)

The problem for a registered user does not occur. 
![image](https://user-images.githubusercontent.com/109066311/201581761-1eeef5ed-dc4a-4276-b148-3a48de9fd88d.png)

**Comment by btry** [comment]
Hi

I reproduced. The plugin works with GLPI 10.0.3, but breaks again with GLPI 10.0.5 (or 10.0.4).

**Comment by Gitnor77** [comment]
Ok, I'll be waiting.
The same problem with GLPI 10.0.4.

**Comment by Gitnor77** [comment]
Hi,
maybe update?

**Comment by btry** [comment]
Hi

Unfortunately not. I have a lot of work and don't have time to work on community issues. 

Community support is at best effort, and no customer seems to have this issue so far.

Also, I plan to release the version 2.13.3 (initially this week, but may be a bit late). This issue will not be fixed in this incoming release.

**Comment by cersyl** [comment]
Bonsoir,

Même problème que mon ticket #3008. Juste pour dire que cela se reproduit suite la MAJ de GLPI 10.0.5 et formcreator 2.13.3, si ca peut aider.
Merci cependant pour le développement de ce superbe plugin !!

**Comment by btry** [comment]
#3095 might be a duplicate

**Comment by willysaurio** [comment]
I have the same problem.
**GLPI 10.0.5 Formcreator 2.13.3**
When attaching a file in public form (anonymous user) the information window appears with the message: **{"message":"The action you have requested is not allowed."}**

**Comment by btry** [comment]
Hi

Please try the fix #3097:
-apply the fix
-upgrade the plugin from CLI with the command bin/console plugin:install formcreator -f -p force-upgrade -p skip-db-check

**Comment by willysaurio** [comment]
Thanks for answering btry.
I do not program in php and I would not know what to modify with respect to fix #3097.
On the other hand the **command bin/console plugin:install formcreator -f -p force-upgrade -p skip-db-check** when executed (in a test GLPI that I have) requires a User which is not understood.
What I would like, like all of us who are reporting this problem, is that the plugin works as it always did allowing unregistered users to generate their claims through public forms and attach, for example, images of the problem they have or other files.
Thank you very much for your time and dedication

**Comment by btry** [comment]
> I do not program in php and I would not know what to modify with respect to fix https://github.com/pluginsGLPI/formcreator/pull/3097.

To apply a patch yoou need to download it by appending .diff at the end of the URL, then use the patch tool to merge the changes in files of a project (see https://man7.org/linux/man-pages/man1/patch.1.html)

> the command bin/console plugin:install formcreator -f -p force-upgrade -p skip-db-check when executed (in a test GLPI that I have) requires a User which is not understood.

you need to specify a super admin user (usually glpi)

> What I would like, like all of us who are reporting this problem, is that the plugin works as it always did

The patch should address the issue, and **it works for me**. I need a confirmation that it works for you or some one else, before merging the fix in the code then make it available in the next release. Also, having a working fix allows you and other people to apply the patch early to not wait the next version.

As you seem around here, please get this build of the plugin. It contains all bugfixes done so far + the above patch. Create a testing instance of GLPI by copying your production. Be careful tu use a **copy** of your database ! Next upgrade the plugin as I described above, then test if the file uploads works well.

[formcreator_2.23.4-dev.zip](https://github.com/pluginsGLPI/formcreator/files/10166074/formcreator_2.23.4-dev.zip)

**Comment by willysaurio** [comment]
Nice day for all

I'm not sure how to use the **patch** tool so I did it differently.

-I moved to another place the formcreator folder that is under .. /marketplace/formcreator
-Lower and unzip the file **formcreator_2.23.4-dev.zip** leaving the same under .. /marketplace/formcreator
-I ran command **bin/console plugin:install formcreator -f -p force-upgrade -p skip-db-check** and assigned a user with super-admin profile
-Check in installed plugins and the plugin appears in version 2.13.4-dev 
-Activated the formcreator plugin
-I made sure I was not logged into the system and ran a form that allows you to attach files.
-The error occurs again but the first time. When you exit the browser and enter again it works and the ticket can be generated with the attachment from the public form with anonymous user. When you opened the ticket it was generated correctly with the attachment inside it.

One more observation
The message that appears when submitting the form is always in English. (**The form has been successfully saved**). Here it is spoken in Spanish and GLPI is set in Spanish-Argentina. Formcreator does not take language into account.

**Comment by btry** [comment]
Hi

>  One more observation
The message that appears when submitting the form is always in English. (The form has been successfully saved). Here it is spoken in Spanish and GLPI is set in Spanish-Argentina. Formcreator does not take language into account.

You were using an public form. In this case the plugin tries to detect the best language for the user. Maybe it failed to find Spanish. 

> The error occurs again but the first time. When you exit the browser and enter again it works and the ticket can be generated with the attachment from the public form with anonymous user. When you opened the ticket it was generated correctly with the attachment inside it.

Can you try again to submit answers and use an other browser (or private mode) to login and check if the ticket is properly created ?

**Comment by willysaurio** [comment]
Good afternoon
I did the tests in private mode and the result is the same.
What I have detected as different is that if I create a "new" public form with the attachment field it works the first time. After that the old forms work without problems.
If you go to another computer and try to run an old form first, the error recurs.

It is as if the "new" form leaves a trace or cookie on the computer which is then used by the other forms. I don't know how to explain it any other way.

If the new unpublished version of the Formcreator 2.13.4 plugin works for the rest of the people who have reported the problem I think you should publish so that we can update from the normal way.

Thanks **btry**

**Comment by willysaurio** [comment]
Nice day.
I think I have found the reason why old public forms generated error when attaching files with the formcreator plugin 2.13.4-dev.

Our entity structure comprises one root entity and several child subentities.

The public forms we have are, according to the destination entity, located in the subentity to which they belong. 
I noticed in the table **glpi_plugin_formcreator_forms** that the new form had 0 as an **entities_id** and the old ones another number, according to the subentity in which they are located.
I transferred one of these forms to the root entity and the problem of attaching files was fixed.

In short, the formcreator 2.13.4-dev plugin does not work for attaching files when the public form is in an entity other than the root.

Hopefully this will help to solve this problem that we are suffering with the formcreator plugin in GLPI 10.0.x

Best regards

**Comment by btry** [comment]
This sounds interesting. Thank you for this feedback

**Comment by Gitnor77** [comment]
> In short, the formcreator 2.13.4-dev plugin does not work for attaching files when the public form is in an entity other than the root.

I confirm, I tested earlier on the main unit.

**Comment by btry** [comment]
Hi
@willysaurio : I thikn I found the problem. I want to check if you have a configuration similar to what I observed.

Please Fund the user with the name beginning with **formcreator_public_user_**. Check his authorizations. He must have rights on the root entity (ID=0) with recursion with the profile named **formcreator_public_profile**.

If you have a configuration different from the above then you need delete the user, and the profile. Ensure that the user is **not** in the trash bin.

Next execute the following SQL query: 
```sql
DELETE FROM glpi_configs WHERE context='formcreator' AND `name`='public_user_id';
```

Finally remove the patch, download it again (to get the latest version) then applyt it.

**Comment by To0tic** [comment]
Good evening,
Good news, it's seems to be good.
I have delete the user and profile, erase the precedent fix and I have reinstall Form creator like your indications.
Thank you very much for your help

**Comment by btry** [comment]
Great ! I'll wait a bit to get more feedbacks on other instances. I expect to have enough positive results next week and release in January (end of year incoming). 


Tha K you for your fast reply.

**Comment by btry** [comment]
Hi

The GLPI team submitted an other solution which works fine with less code. You should revert the patch and apply this one on GLPI: https://github.com/glpi-project/glpi/pull/13641

**Comment by willysaurio** [comment]
@willysaurio : I think I found the problem. I want to check if you have a configuration similar to what I observed.

Please Fund the user with the name beginning with formcreator_public_user_. Check his authorizations. He must have rights on the root entity (ID=0) with recursion with the profile named formcreator_public_profile.

If you have a configuration different from the above then you need delete the user, and the profile. Ensure that the user is not in the trash bin.

**>>>I found the user formcreator_public_user_ and their profile is formcreator_public_profile Non-Recursive.**

Next execute the following SQL query:

DELETE FROM glpi_configs WHERE context='formcreator' AND `name`='public_user_id';

**>>>I didn't do it yet because I don't quite understand the next steps.**

Remove the patch, 

**>>> I must uninstall from the same GLPI system or remove .../marketplace/formcreator from the console?
>>> In case it is the first all the forms created will be lost because uninstalling the tables of formcreator in the database are deleted....**

download it again (to get the latest version) then applyt it.
**>>> Could you please indicate the link to the version of formcreator you refer?
>>> Where the referred version should be downloaded. In.. /glpi/plugins/formcreator or in .. /glpi/marketplace/formcreator ?**

**Comment by btry** [comment]
Hi

An other solution has been tested. Remove the patch on foemcreator, and apply this one on glpi. We had positive feedback for now and it is simpler.

https://github.com/glpi-project/glpi/pull/13641

**Comment by btry** [comment]
Oh sorry. I thought this was an issue where I did not submitted the patch for glpi.

If you do 't know how to execute the SQL query, you can  it ore this step. The data, if not deleted, is harmless, it will just be garbage.

**Comment by willysaurio** [comment]
> Hi
> 
> An other solution has been tested. Remove the patch on foemcreator, and apply this one on glpi. We had positive feedback for now and it is simpler.
> 
> [glpi-project/glpi#13641](https://github.com/glpi-project/glpi/pull/13641)

I'm confused. I can't find the download link of the plugin you refer. Could you write the download link of the formcreator plugin that you say should be installed?.

**Comment by willysaurio** [comment]
> Oh sorry. I thought this was an issue where I did not submitted the patch for glpi.
> 
> If you do 't know how to execute the SQL query, you can ignore this step. The data, if not deleted, is harmless, it will just be garbage.

I know how to execute the SQL query but i´m confused about de nexts steps

**Comment by willysaurio** [comment]
I'm not an expert on linux. I am just an administrator of the GLPI tool.

**Comment by btry** [comment]
Ok. Here is the download link of Formcreator 2.13.3 

https://github.com/pluginsGLPI/formcreator/releases/download/2.13.3/glpi-formcreator-2.13.3.tar.bz2

Check in GLPI, folders plugins/ and marketplace/. Only one of them shall contain the plugin. Remove the subfolder formcreator from the place you found it (move it away from glpi), then use the above link to install the latest version.

Next, apply the patch for GLPI

**Comment by willysaurio** [comment]
> To remove a patch you need to apply it again with the patch tool, with the extra argument -R . It means "reverse", then the tool will remove the patch, instead of add it.
> 
> As you seem a bit lost, you may also install an unpatched formcreator copy in the filesystem.

The user formcreator_public_user_ is **Non-Recursive** and their profile is formcreator_public_profile 
- I apply DELETE FROM glpi_configs WHERE context='formcreator' AND name='public_user_id'

- Downloaded https://github.com/pluginsGLPI/formcreator/releases/download/2.13.3/glpi-formcreator-2.13.3.tar.bz2.
- I moved - ..marketplace/formcreator to another place.
- I unzipped glpi-formcreator-2.13.3.tar.bz2 
- I moved the formcreartor (2.13.3) floder to ..marketplace/forcreator

What should I do now?:

bin/console plugin:install formcreator -u <super adm usr> -f -p force-upgrade -p skip-db-check
or 
bin/console plugin:install formcreator -u <super adm usr> -f -p force-upgrade -p skip-db-check -R

**Comment by btry** [comment]
This one

bin/console plugin:install formcreator -u -f -p force-upgrade -p skip-db-check

**Comment by willysaurio** [comment]
Apply **bin/console plugin:install formcreator -u <Usr super adm>
![bin-console](https://user-images.githubusercontent.com/65296683/208439533-81599703-b6d1-4726-b5c4-c643a774b470.jpg)
 -f -p force-upgrade -p skip-db-check**
I get this result in console:

**Comment by btry** [comment]
Did you apply the patch for glpi ?

https://github.com/glpi-project/glpi/pull/13641

**Comment by willysaurio** [comment]
No. I also don't know where to look for the patch you mention. Please pass me the patch link.

**Comment by btry** [comment]
Click on the blue link.

To download the raw patch file, use curl or wget and add .diff to the URL. 

Finally use patch tool to apply it to glpi. There are plenty of resources on internet to learn how to use the tool. You also may ask on the forum dedicated to GLPI.

**Comment by willysaurio** [comment]
Is the patch just modifying the "src/Session.php" file or something else?

**Comment by btry** [comment]
It changes several files. Don't change files manually. Always use the patch tool.

**Comment by willysaurio** [comment]
"Patch" is very complex for me to understand.
Can I use rsync instead of patch to update the differences?
( rsync -r  /source /destination)

If it is not possible I will try to find someone to explain to me how it should be used. (My English is limited and I don't understand some things.) Thank you

**Comment by btry** [comment]
Do not use rsync. 

I assume that glpi is in /var/www/html/glpi

cd $HOME
wget https://github.com/glpi-project/glpi/pull/13641.diff
cd /var/www/html/glpi 
patch -p1 --dry-run < $HOME/13641.diff

On any error, ABORT and show the output of the command.

If the command is successful, repeat it without the argument --dry-run

**Comment by btry** [comment]
Last thing. 

If you need to remove the patch, run
patch -p1 -R < $HOME/13641.diff

**Comment by btry** [comment]
Yes, this is ok

**Comment by willysaurio** [comment]
> Yes, this is ok
I ran the command without --dry-run and then tested.
The patch of the src/Session.php, ajax/fileupload.php and ajax/getFileTag.php files worked correctly with formcreator version 2.13.3
I was able to create tickets from public forms by attaching a file seamlessly.
Thank you very much for your willingness and patience.

**Comment by btry** [comment]
Great. 

You can now delete the user and the profile created by the previous match of Formcreator.

I'll ask the glpi team to merge the fix, so it will hopefully be available in the next release

**Comment by cersyl** [comment]
The patch of the src/Session.php, ajax/fileupload.php and ajax/getFileTag.php files worked correctly with formcreator version 2.13.3 for me too :-)

Thank you very much for your great job !!!

**Comment by btry** [comment]
Hi

Thank you for your feedback. I think I got enough feedbacks to validate the fix. I close this issue.
